migrations/2023-08-01-115243_persistent-activity-queue/up.sql

```diff
@@ -0,0 +1,31 @@
+CREATE TYPE actor_type_enum AS enum(
+    'site',
+    'community',
+    'person'
+);
+
+-- actor_apub_id only null for old entries before this migration
+ALTER TABLE sent_activity
+    ADD COLUMN send_inboxes text[] NOT NULL DEFAULT '{}', -- list of specific inbox urls
```

Comment:
Probably the biggest no-no in SQL, is to avoid storing arrays in columns. If you ever need to query this table and filter on this column, you won't be able to do it smartly with SQL. As long as you're 100% positive you'll never need to filter on this column, then I spose its okay. But otherwise I would properly normalize this and break this out into a bridge table, and use joins. Postgres should be doing the filtering / selecting, not rust.

Reply:
i know that in general you don't want arrays in a sql db. my older version actually had a jsonb column that had more other stuff, but i assumed you would like that even less.

the thing is that i thought of different approaches of how to do this, and there is no way to make any kind of db filtering here be useful.

* if you actually put all the instances into the db, the db load would increase by 100-1000x because you need one row per activity per receiving instance.

* if you do the filtering within the db, then you need again, one differing query per instance, and you cannot take advantage of caching at all.

obviously i'm open to different ways of doing this but i have not heard any better approaches so far.

Reply:
i have replaced the community_followers_of array with a single nullable value because there can't ever actually be multiple of those. send_inboxes is still an array though. it's only used for comment mentions (currently) and is mostly empty

Reply:
I spose I'm fine with this, but I don't see why it would balloon the DB so much. The sent activities get periodically cleared, and all the connected send_inbox rows would get deleted with them.

Where you extend the inbox_urls with the send_inbox rows could be filtered by `instance_id`, or the domain, and I doubt it'd require looping queries or anything

But you don't have to do it for this one.